Fix GIL=0 segfault and Add GIL=0 compat for regex paths

Qubitium opened the pull request:
# What does this PR do?

Adds (Full) Regex and (Partial) Tokenization GIL=0 free-threading support. Tested up to Python 3.14T.

In simple terms, Transformers code that relies on regex will segfault under true concurrency. I have confirmed with the regex maintainer that the latest GIL=0–compatible regex package is not GIL=0 safe. This has been corroborated by the submitted unit test in this PR, which demonstrates the segfault. Regex caches and reuses compiled patterns internally and executes them, which makes it inherently unsafe without the GIL.

The core idea is not to make the entire Transformers library GIL=0 safe all at once, but rather to adapt it piece by piece until everything is as GIL=0 compatible as possible.

This PR wraps existing regex calls into a thread-locked, protected serial execution pipeline. Files that import regex only need to adjust their import path to `from util.safe import regex` to minimize migration pain.

```diff
- import regex as re
+ from ...utils.safe import regex as re
```

The current safe wrapper is modular and can be extended to cover additional modules in the future as needed or as issues are discovered. 

Since Transformers evolves more rapidly than most packages, introducing a local wrapper may be a pragmatic step while waiting for an upstream fix, which may be delayed. Additionally, many if not most Python APIs are not inherently thread-safe, so applying local wrappers is often necessary beyond just external packages.”

Two new unit test files are included:

A non-crashing test suite that proves the code works under GIL=0 with thread load.

A crash-demonstration test that proves regex code paths segfault without the added protection under thread load.

[utils/safe.py](https://github.com/huggingface/transformers/compare/main...Qubitium:transformers:regex-thread-safety?expand=1#diff-a7a62ce35afa10519218152098da57fe8ad5012b0068cfcb8e7ced98084e1517)

[tests/utils/test_safe.py](https://github.com/huggingface/transformers/compare/main...Qubitium:transformers:regex-thread-safety?expand=1#diff-c36421043b36005fb56f58552a69701e7694997a64c1083816b888494f2666ac)

[tests/utils/test_safe_crash.py](https://github.com/huggingface/transformers/compare/main...Qubitium:transformers:regex-thread-safety?expand=1#diff-2c867bbc58cad710595119fe34ae6389ba7f088f799714b2e7c6e93e72893bc1)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

@ArthurZucker @itazap @SunMarc @Cyrilvallez @gante @ydshieh  @stevhliu

Qubitium committed:
Fix GIL=0 segfault and Add GIL=0 compat for regex/tokenization

Qubitium committed:
Merge branch 'main' into regex-thread-safety

Rocketknight1 reviewed (commented):
This seems like a really cool idea! I made one comment about `locked`, but it makes a lot of sense otherwise.

One potential simplification, though, is that in a lot of cases we `import regex` when really we only need `re`; this is a leftover from a time when the built-in `re` was significantly behind the third-party `regex` lib. What's the thread-safety status of the built-in `re`?

Qubitium committed:
Merge branch 'main' into regex-thread-safety

Qubitium commented:
> This seems like a really cool idea! I made one comment about `locked`, but it makes a lot of sense otherwise.
> 
> One potential simplification, though, is that in a lot of cases we `import regex` when really we only need `re`; this is a leftover from a time when the built-in `re` was significantly behind the third-party `regex` lib. What's the thread-safety status of the built-in `re`?

Another issue about`re` is it's `api` is highly unstable. There are breaking changes as recent as 3.12 and I just took a casual glance at the doc: https://docs.python.org/3/library/re.html. 

From docs `re` also cache compiled regex so I think we can assume, without look at the implementation, that it may well also be thread unsafe. Need to double check with internal code. 
```
Note The compiled versions of the most recent patterns passed to [re.compile()](https://docs.python.org/3/library/re.html#re.compile) and the module-level matching functions are cached, so programs that use only a few regular expressions at a time needn’t worry about compiling regular expressions.
```

```
Changed in version 3.12: In [bytes](https://docs.python.org/3/library/stdtypes.html#bytes) patterns, group name can only contain bytes in the ASCII range (b'\x00'-b'\x7f').
```

```
Changed in version 3.12: Group id can only contain ASCII digits. In [bytes](https://docs.python.org/3/library/stdtypes.html#bytes) patterns, group name can only contain bytes in the ASCII range (b'\x00'-b'\x7f').
```

Qubitium committed:
Merge branch 'main' into regex-thread-safety

ArthurZucker reviewed (commented):
Hey! Nice PR, we are removing all of these files in favor of just using `tokenizers` as the backend, and having one `sentencepiece_wrapper` file, from which we will use your safe import!

Qubitium committed:
Fix `_callable_cache` not lock protected

Qubitium committed:
Fix some python modules (pytorch) do not have `__package__` meta field

Qubitium committed:
Fix namespace collision with original object with `_hf_safe_` prefix for all properties/helpers

Qubitium committed:
Merge branch 'main' into regex-thread-safety

Qubitium committed:
Clean comments

Qubitium committed:
Fix reentrant usage cases and add regex reentrant test case

Qubitium committed:
Merge branch 'main' into regex-thread-safety

Qubitium committed:
Merge branch 'main' into regex-thread-safety

Qubitium committed:
Merge branch 'main' into regex-thread-safety

Qubitium committed:
Merge branch 'main' into regex-thread-safety

Qubitium committed:
Merge branch 'main' into regex-thread-safety

Rocketknight1 commented on src/transformers/utils/safe.py:
```diff
@@ -0,0 +1,121 @@
+            @wraps(attr)
+            def locked(*args, **kwargs):
+                with self._lock:
+                    return attr(*args, **kwargs)
```
I can't think of an obvious one, unfortunately! You could make it less visible in the code by defining the attribute name at top of the module as a constant, and then just referencing the constant :sweat_smile:

Rocketknight1 commented:
Got it! It seems good now, but can we confirm the status of the built-in `re` in `3.14T`? If we can avoid thread-safety wrappers by just swapping to that, I think I'd prefer it, even if there are minor API changes between versions. I don't think we should be affected by those since I doubt we have non-ASCII capture group names.

Of course, if the built-in `re` has the same thread-safety issues, then we should definitely go with something like this PR

Qubitium committed:
Merge branch 'main' into regex-thread-safety

Qubitium committed:
update tests for re.match and pcre2

Qubitium committed:
Merge branch 'regex-thread-safety' of https://github.com/Qubitium/transformers into regex-thread-safety

Qubitium committed:
format

Qubitium committed:
remove pcre2 and add re.match

Qubitium commented:
@Rocketknight1 I believe move to `re` should be on another PR (asap due to result from following benchmark) as it would need validate all the existing regex is input/output compatible but `re` is the future here. `pcre2` with JIT is good but can't beat `re` in mult-threaded at least for this simple bench. 

I have tested `re` is thead-safe and very performant even compared with `pcre2`.

As for benchmarking, `regex` vs `safe.regex` is on par.  `safe`'s wrapper + extra lock introduced minor performance degradation on singlethread.

* `re` is thread-safe
* `regex` is 10x slower than `re`. This was unexpected. 
* `pcre2` is only faster than `re` in singler threaded JIT-enabled mode. 

Benchamark only tested 2 regex match: simple + named grouped matching. 

Conclusion Transformers should move to `re` for any tokenization related code. 

```py
(vm314t) root@gpu-base:~/transformers# PYTHON_GIL=0 python benchmark_re_vs_regex.py
Single-thread loops per matcher: 10000
Threaded loops: 8 threads x 10000 loops (PYTHON_GIL=0)

Pattern case: simple
  expr: (Transformers|transformers|models|Models)


Single-thread
Library        Time (s)  Ops/s        Matches    Total Ops
-----------  ----------  ---------  ---------  -----------
re               0.0063  1,580,989      10000        10000
regex            0.0605  165,181        10000        10000
safe.regex       0.065   153,862        10000        10000
pcre2            0.0151  660,429        10000        10000
pcre2 (jit)      0.0177  564,149        10000        10000


Threaded
Library        Time (s)  Ops/s      Matches    Total Ops
-----------  ----------  -------  ---------  -----------
re               0.0865  925,350      80000        80000
safe.regex       1.7687  45,231       80000        80000
pcre2            0.1264  633,120      80000        80000
pcre2 (jit)      0.126   635,079      80000        80000

Pattern case: named-groups
  expr: (?P<head>\w+)\s+(?P<tail>\w+)


Single-thread
Library        Time (s)  Ops/s      Matches    Total Ops
-----------  ----------  -------  ---------  -----------
re               0.0188  531,530      10000        10000
regex            0.0749  133,589      10000        10000
safe.regex       0.0633  157,934      10000        10000
pcre2            0.0196  508,918      10000        10000
pcre2 (jit)      0.0149  669,002      10000        10000


Threaded
Library        Time (s)  Ops/s      Matches    Total Ops
-----------  ----------  -------  ---------  -----------
re               0.1135  704,786      80000        80000
safe.regex       1.8469  43,317       80000        80000
pcre2            0.1391  575,179      80000        80000
pcre2 (jit)      0.1366  585,715      80000        80000
```

Benchmar code:

```py
#!/usr/bin/env python
"""Micro benchmark comparing stdlib re with transformers.utils.safe.regex and pcre2"""

import os
import re
import threading
import time
from typing import Callable, Sequence, Tuple

from tabulate import tabulate

from transformers.utils.safe import regex as safe_regex
import regex
import pcre2

SINGLE_THREAD_LOOPS = 10_000
THREAD_LOOPS = 10_000
THREAD_COUNT = 8
SIMPLE_PATTERN = r"(Transformers|transformers|models|Models)"
SIMPLE_TEXT = "Transformers by Hugging Face enable models across tasks and devices."

COMPLEX_PATTERN = r"(?P<head>\w+)\s+(?P<tail>\w+)"
COMPLEX_TEXT = "Transformers models"


def string_matcher_factory(
    match_func: Callable[[str, str], object]
) -> Callable[[str, str], Callable[[], bool]]:
    def factory(pattern: str, text: str) -> Callable[[], bool]:
        def _match() -> bool:
            return match_func(pattern, text) is not None

        return _match

    return factory


def pcre2_compiled_factory(jit: bool) -> Callable[[str, str], Callable[[], bool]]:
    def factory(pattern: str, text: str) -> Callable[[], bool]:
        compiled = _get_pcre2_pattern(pattern, jit=jit)
        match = compiled.match

        def _match() -> bool:
            return match(text) is not None

        return _match

    return factory


def benchmark_single(label: str, match_callable: Callable[[], bool]):
    start = time.perf_counter()
    matches = 0
    for _ in range(SINGLE_THREAD_LOOPS):
        matches += match_callable()
    duration = time.perf_counter() - start
    ops_per_sec = SINGLE_THREAD_LOOPS / duration if duration else float("inf")
    return duration, ops_per_sec, matches


def benchmark_threaded(label: str, match_callable: Callable[[], bool]):
    total_matches = [0] * THREAD_COUNT

    def worker(slot: int) -> None:
        local_match = match_callable
        count = 0
        for _ in range(THREAD_LOOPS):
            count += local_match()
        total_matches[slot] = count

    threads = [threading.Thread(target=worker, args=(idx,), name=f"{label}-worker-{idx}") for idx in range(THREAD_COUNT)]
    start = time.perf_counter()
    for thread in threads:
        thread.start()
    for thread in threads:
        thread.join()
    duration = time.perf_counter() - start
    total_ops = THREAD_COUNT * THREAD_LOOPS
    total = sum(total_matches)
    ops_per_sec = total_ops / duration if duration else float("inf")
    return duration, ops_per_sec, total

MatcherFactory = Callable[[str, str], Callable[[], bool]]


_PCRE2_CACHE: dict[tuple[str, bool], pcre2.Pattern] = {}
_PCRE2_CACHE_LOCK = threading.Lock()


def _get_pcre2_pattern(pattern: str, *, jit: bool) -> pcre2.Pattern:
    key = (pattern, jit)
    cached = _PCRE2_CACHE.get(key)
    if cached is not None:
        return cached

    with _PCRE2_CACHE_LOCK:
        cached = _PCRE2_CACHE.get(key)
        if cached is None:
            cached = pcre2.compile(pattern, jit=jit)
            _PCRE2_CACHE[key] = cached
        return cached


SINGLE_THREAD_LIBRARIES: Sequence[Tuple[str, MatcherFactory]] = (
    ("re", string_matcher_factory(re.match)),
    ("regex", string_matcher_factory(regex.match)),
    ("safe.regex", string_matcher_factory(safe_regex.match)),
    ("pcre2", pcre2_compiled_factory(jit=False)),
    ("pcre2 (jit)", pcre2_compiled_factory(jit=True)),
)


THREADED_LIBRARIES: Sequence[Tuple[str, MatcherFactory]] = (
    ("re", string_matcher_factory(re.match)),
    ("safe.regex", string_matcher_factory(safe_regex.match)),
    ("pcre2", pcre2_compiled_factory(jit=False)),
    ("pcre2 (jit)", pcre2_compiled_factory(jit=True)),
)


PATTERN_CASES: Sequence[Tuple[str, str, str]] = (
    ("simple", SIMPLE_PATTERN, SIMPLE_TEXT),
    ("named-groups", COMPLEX_PATTERN, COMPLEX_TEXT),
)


if __name__ == "__main__":
    print(f"Single-thread loops per matcher: {SINGLE_THREAD_LOOPS}")
    total_thread_ops = THREAD_COUNT * THREAD_LOOPS
    os.environ["PYTHON_GIL"] = "0"
    print(
        f"Threaded loops: {THREAD_COUNT} threads x {THREAD_LOOPS} loops (PYTHON_GIL={os.environ['PYTHON_GIL']})"
    )

    for case_label, pattern, text in PATTERN_CASES:
        print()
        print(f"Pattern case: {case_label}")
        print(f"  expr: {pattern}")

        single_rows = []
        for label, matcher_factory in SINGLE_THREAD_LIBRARIES:
            duration, ops_per_sec, matches = benchmark_single(
                label, matcher_factory(pattern, text)
            )
            single_rows.append(
                [label, f"{duration:.4f}", f"{ops_per_sec:,.0f}", matches, SINGLE_THREAD_LOOPS]
            )
        print("\n\nSingle-thread")
        print(
            tabulate(
                single_rows,
                headers=["Library", "Time (s)", "Ops/s", "Matches", "Total Ops"],
            )
        )

        threaded_rows = []
        for label, matcher_factory in THREADED_LIBRARIES:
            duration, ops_per_sec, matches = benchmark_threaded(
                label, matcher_factory(pattern, text)
            )
            threaded_rows.append(
                [
                    label,
                    f"{duration:.4f}",
                    f"{ops_per_sec:,.0f}",
                    matches,
                    total_thread_ops,
                ]
            )
        print("\n\nThreaded")
        print(
            tabulate(
                threaded_rows,
                headers=["Library", "Time (s)", "Ops/s", "Matches", "Total Ops"],
            )
        )
```

Qubitium committed:
update test

Qubitium committed:
Merge branch 'main' into regex-thread-safety

Qubitium commented:
@Rocketknight1  @ArthurZucker While testing `regex` -> `re` replacement I find that the tokenicers use `\p{L}` language group matching which is not supported by `stdlib.re`. This is a huge compat issue. So based on my benchmark, the future is not `regex` -> `re` but `regex` -> `pcre2`.

Qubitium commented:
@Rocketknight1 @ArthurZucker Existing tokenizer code use `\p{L}` language group modifiers for matching which `stdlb.re` does not yet support so `re` is out of the question. So the future question is should Transofmers migrate to `pcre2` instead which matches `re` in threaded speed with full pcre compatible syntax support.

```py
# tokenization_qwen2.py
import regex as re

from ...tokenization_utils import AddedToken, PreTrainedTokenizer
from ...utils import logging 


logger = logging.get_logger(__name__)

VOCAB_FILES_NAMES = {
    "vocab_file": "vocab.json",
    "merges_file": "merges.txt",
}


MAX_MODEL_INPUT_SIZES = {"qwen/qwen-tokenizer": 32768}

PRETOKENIZE_REGEX = r"""(?i:'s|'t|'re|'ve|'m|'ll|'d)|[^\r\n\p{L}\p{N}]?\p{L}+|\p{N}| ?[^\s\p{L}\p{N}]+[\r\n]*|\s*[\r\n]+|\s+(?!\S)|\s+"""
```

Qubitium committed:
update test to show regex is not thread-safe even for non-cached pattern

Qubitium commented on tests/utils/test_safe_crash.py:
```diff
@@ -114,6 +114,33 @@ def match_once():
     pytest.fail(message)
 
 
+@pytest.mark.xfail(strict=False, reason="Compiled regex still crashes under PYTHON_GIL=0")
+def test_compiled_regex_thread_safety_crashes_under_gil0():
+    script = _regex_thread_script(
+        imports="import regex",
+        setup_code="""
+        compiled_pattern = regex.compile(pattern_text)
+
+        def match_once():
+            return compiled_pattern.match(text_to_match)
+        """,
+    )
+
+    result, message = _run_regex_thread_script("compiled regex", script)
+
+    if result.returncode == 0:
+        pytest.fail("compiled regex unexpectedly behaved thread-safely\n" + message)
+
+    if result.returncode == -11:
+        message += "\nProcess terminated with SIGSEGV (Segmentation fault)."
+
+    if message:
+        sys.stderr.write(message + "\n")
+        sys.stderr.flush()
+
+    pytest.fail(message)
+
+
```
Add new test to show `regex` is crashing with GIL=0 even when the pattern is not going through the caching mechanism which I thought was the cause of the previous crashes. This is strange and unexpected. It appears `regex` lib is fully unsafe as it maybe reusing the same runtime stack/buffers for same pattern execution.

Qubitium committed:
update comment to show that even compiled pattern is not thread safe

Qubitium commented:
Update benchmark adding `python-pcre` comparison. Looks ike `python-pcre` is clear winner. 

```py
Single-thread loops per matcher: 10000
Threaded loops: 8 threads x 10000 loops (PYTHON_GIL=0)

Pattern case: simple
  expr: (Transformers|transformers|models|Models)


Single-thread
Library        Time (s)  Ops/s        Matches    Total Ops
-----------  ----------  ---------  ---------  -----------
re               0.0059  1,684,404      10000        10000
regex            0.0631  158,556        10000        10000
safe.regex       0.0646  154,914        10000        10000
pcre             0.0055  1,829,647      10000        10000
pcre2            0.0148  677,749        10000        10000
pcre2 (jit)      0.0156  642,918        10000        10000


Threaded
Library        Time (s)  Ops/s      Matches    Total Ops
-----------  ----------  -------  ---------  -----------
re               0.2125  376,419      80000        80000
safe.regex       1.7535  45,622       80000        80000
pcre             0.093   860,518      80000        80000
pcre2            0.2944  271,753      80000        80000
pcre2 (jit)      0.3436  232,837      80000        80000

Pattern case: named-groups
  expr: (?P<head>\w+)\s+(?P<tail>\w+)


Single-thread
Library        Time (s)  Ops/s        Matches    Total Ops
-----------  ----------  ---------  ---------  -----------
re               0.0164  608,513        10000        10000
regex            0.0908  110,083        10000        10000
safe.regex       0.0625  160,029        10000        10000
pcre             0.0055  1,822,747      10000        10000
pcre2            0.0149  670,923        10000        10000
pcre2 (jit)      0.0148  673,560        10000        10000


Threaded
Library        Time (s)  Ops/s      Matches    Total Ops
-----------  ----------  -------  ---------  -----------
re               0.2471  323,816      80000        80000
safe.regex       1.8391  43,499       80000        80000
pcre             0.0994  805,034      80000        80000
pcre2            0.3448  231,985      80000        80000
pcre2 (jit)      0.3114  256,917      80000        80000
(vm314t) root@gpu-base:~/transformers# PYTHON_GIL=0 python benchmark*py
Single-thread loops per matcher: 10000
Threaded loops: 8 threads x 10000 loops (PYTHON_GIL=0)

Pattern case: simple
  expr: (Transformers|transformers|models|Models)


Single-thread
Library        Time (s)  Ops/s        Matches    Total Ops
-----------  ----------  ---------  ---------  -----------
re               0.0059  1,683,494      10000        10000
regex            0.069   144,929        10000        10000
safe.regex       0.0623  160,490        10000        10000
pcre             0.0055  1,827,192      10000        10000
pcre (jit)       0.0057  1,759,222      10000        10000
pcre2            0.0159  629,413        10000        10000
pcre2 (jit)      0.0145  690,844        10000        10000


Threaded
Library        Time (s)  Ops/s        Matches    Total Ops
-----------  ----------  ---------  ---------  -----------
re               0.109   734,058        80000        80000
safe.regex       1.7284  46,286         80000        80000
pcre             0.0528  1,516,316      80000        80000
pcre (jit)       0.0544  1,471,912      80000        80000
pcre2            0.1472  543,591        80000        80000
pcre2 (jit)      0.1346  594,521        80000        80000

Pattern case: named-groups
  expr: (?P<head>\w+)\s+(?P<tail>\w+)


Single-thread
Library        Time (s)  Ops/s        Matches    Total Ops
-----------  ----------  ---------  ---------  -----------
re               0.0193  517,076        10000        10000
regex            0.0825  121,173        10000        10000
safe.regex       0.0619  161,673        10000        10000
pcre             0.0055  1,815,004      10000        10000
pcre (jit)       0.0064  1,571,113      10000        10000
pcre2            0.0167  598,186        10000        10000
pcre2 (jit)      0.0153  653,871        10000        10000


Threaded
Library        Time (s)  Ops/s        Matches    Total Ops
-----------  ----------  ---------  ---------  -----------
re               0.1167  685,416        80000        80000
safe.regex       1.7527  45,643         80000        80000
pcre             0.0528  1,515,403      80000        80000
pcre (jit)       0.0517  1,546,954      80000        80000
pcre2            0.123   650,669        80000        80000
pcre2 (jit)      0.1326  603,108        80000        80000
```

Qubitium committed:
Merge branch 'main' into regex-thread-safety

Rocketknight1 commented:
Interesting, and thank you for the comprehensive benchmarks! I'd have to discuss with the team, but I suspect our stance is that we want to support free-threading, but not at the cost of other PRs that have a high chance of breaking things. If we can do a drop-in replacement for `regex` then I think we can support that, but if we have to move to a new regex engine and start changing patterns across the codebase then we might have to just wait for patches to `regex` itself.

That said, regex performance is usually not a bottleneck for tokenization, let alone LLM inference in general, so if `pcre2` works perfectly but is just a little slower than `regex` or `re` then that's probably fine!

Qubitium committed:
Merge branch 'main' into regex-thread-safety

Qubitium commented:
> Interesting, and thank you for the comprehensive benchmarks! I'd have to discuss with the team, but I suspect our stance is that we want to support free-threading, but not at the cost of other PRs that have a high chance of breaking things. If we can do a drop-in replacement for `regex` then I think we can support that, but if we have to move to a new regex engine and start changing patterns across the codebase then we might have to just wait for patches to `regex` itself.
> 
> That said, regex performance is usually not a bottleneck for tokenization, let alone LLM inference in general, so if `pcre2` works perfectly but is just a little slower than `regex` or `re` then that's probably fine!

Agreed. That's for a future (potential) PR and can be discussed with more detail later. I actually went over all the pcre wrappers and/or compat regex engine for python and each has problems or issues that I want to address in my own pcre pkg. They either are not maintained, compile pcre from source (ouch) and/or have usability/compat or GIL=0 issues. 

Let's just return our focus to this simpler PR instead which has no compat issues but inherit all the slowness of `regex`.

Qubitium committed:
Merge branch 'main' into regex-thread-safety

Qubitium committed:
Fix imports relative depth for deprecated tokenizers

Qubitium committed:
Merge branch 'huggingface:main' into regex-thread-safety

Qubitium committed:
Merge branch 'main' into regex-thread-safety

Qubitium committed:
Merge branch 'main' into regex-thread-safety

Qubitium committed:
Merge branch 'main' into regex-thread-safety

Qubitium committed:
Merge branch 'main' into regex-thread-safety

Qubitium committed:
Merge branch 'main' into regex-thread-safety
